### PR TITLE
fix case where an IterableDataset doesn't produce a batch for an epoch

### LINKED ISSUE #6810: Handle cases where an IterableDataset doesn't produce a batch for an epoch.
## 🐛 Bug

If the IterableDataset for training doesn't generate any batch for an epoch, the Trainer raises a rather cryptic exception:
```
Traceback (most recent call last):
  File "/anaconda/envs/py37_pytorch/lib/python3.7/site-packages/pytorch_lightning/trainer/connectors/data_connector.py", line 47, in _with_is_last
    last = next(it)
  File "/anaconda/envs/py37_pytorch/lib/python3.7/site-packages/pytorch_lightning/trainer/supporters.py", line 470, in __next__
    return self.request_next_batch(self.loader_iters)
  File "/anaconda/envs/py37_pytorch/lib/python3.7/site-packages/pytorch_lightning/trainer/supporters.py", line 484, in request_next_batch
    return apply_to_collection(loader_iters, Iterator, next)
  File "/anaconda/envs/py37_pytorch/lib/python3.7/site-packages/pytorch_lightning/utilities/apply_func.py", line 84, in apply_to_collection
    return function(data, *args, **kwargs)
  File "/anaconda/envs/py37_pytorch/lib/python3.7/site-packages/torch/utils/data/dataloader.py", line 435, in __next__
    data = self._next_data()
  File "/anaconda/envs/py37_pytorch/lib/python3.7/site-packages/torch/utils/data/dataloader.py", line 475, in _next_data
    data = self._dataset_fetcher.fetch(index)  # may raise StopIteration
  File "/anaconda/envs/py37_pytorch/lib/python3.7/site-packages/torch/utils/data/_utils/fetch.py", line 34, in fetch
    data = next(self.dataset_iter)
StopIteration

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
...
  File "/anaconda/envs/py37_pytorch/lib/python3.7/site-packages/pytorch_lightning/trainer/trainer.py", line 499, in fit
    self.dispatch()
  File "/anaconda/envs/py37_pytorch/lib/python3.7/site-packages/pytorch_lightning/trainer/trainer.py", line 546, in dispatch
    self.accelerator.start_training(self)
  File "/anaconda/envs/py37_pytorch/lib/python3.7/site-packages/pytorch_lightning/accelerators/accelerator.py", line 73, in start_training
    self.training_type_plugin.start_training(trainer)
  File "/anaconda/envs/py37_pytorch/lib/python3.7/site-packages/pytorch_lightning/plugins/training_type/training_type_plugin.py", line 114, in start_training
    self._results = trainer.run_train()
  File "/anaconda/envs/py37_pytorch/lib/python3.7/site-packages/pytorch_lightning/trainer/trainer.py", line 637, in run_train
    self.train_loop.run_training_epoch()
  File "/anaconda/envs/py37_pytorch/lib/python3.7/site-packages/pytorch_lightning/trainer/training_loop.py", line 484, in run_training_epoch
    for batch_idx, (batch, is_last_batch) in train_dataloader:
  File "/anaconda/envs/py37_pytorch/lib/python3.7/site-packages/pytorch_lightning/profiler/profilers.py", line 82, in profile_iterable
    value = next(iterator)
RuntimeError: generator raised StopIteration

Set the environment variable HYDRA_FULL_ERROR=1 for a complete stack trace.
```
This is caused by [_with_is_last](https://github.com/PyTorchLightning/pytorch-lightning/blob/22a266d8b8cf57455cc863e20491e416ec635ba7/pytorch_lightning/trainer/connectors/data_connector.py#L50-L60) as it assumes the given iterator produces at least one batch. If it doesn't, the first call to [next()](https://github.com/PyTorchLightning/pytorch-lightning/blob/22a266d8b8cf57455cc863e20491e416ec635ba7/pytorch_lightning/trainer/connectors/data_connector.py#L54) raises a `StopIteration`, which gets transformed to a `RuntimeError` per the new behavior in Python 3.7 (https://www.python.org/dev/peps/pep-0479/). 

## Please reproduce using the BoringModel

[Colab link](https://colab.research.google.com/drive/16-wQJPKc0hmOJ2ZqxAAhC7UP4SPTW2PN?usp=sharing)

### To Reproduce

Run the colab.

### Expected behavior

We should do something like the following:
```python
    def _with_is_last(self, iterable):
        """Pass through values from the given iterable with an added boolean indicating if this is the last item.
        See `https://stackoverflow.com/a/1630350 <https://stackoverflow.com/a/1630350>`_"""
        it = iter(iterable)
        try:
            last = next(it)
        except StopIteration:
            warn('your iterable dataset didn't produce any batch')
            return
        for val in it:
            # yield last and has next
            yield last, False
            last = val
        # yield last, no longer has next
        yield last, True
```
If we return like the snippet above, it doesn't quite work b/c the whole training loop code assumes that there's at least one batch. We probably should re-raise the exception with a better message.

### Environment

See the Colab.

### Additional context
N/A

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #6810 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by ethanwharris — APPROVED** [review]
LGTM :smiley:

File: CHANGELOG.md [hunk 1]
```diff
@@ -407,6 +407,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 
 
+- Fixed handling an `IterableDataset` that fails to produce a batch at the beginning of an epoch ([#7294](https://github.com/PyTorchLightning/pytorch-lightning/pull/7294))
```
Comment (kaushikb11) [inline CHANGELOG.md:410]:
```suggestion


- Fixed handling an `IterableDataset` that fails to produce a batch at the beginning of an epoch ([#7294](https://github.com/PyTorchLightning/pytorch-lightning/pull/7294))
```

File: pytorch_lightning/trainer/supporters.py [hunk 2]
```diff
@@ -508,3 +508,25 @@ def _nested_calc_num_data(data: Union[Mapping, Sequence], compute_func: Callable
...
+    """
+    Returns an iterator that pre-fetches and caches the next item.
+    The values are passed through from the given iterable with an added boolean indicating if this is the last item.
+    See `https://stackoverflow.com/a/1630350 <https://stackoverflow.com/a/1630350>`_
+    """
+    it = iter(iterable)
+
+    try:
+        # the iterator may be empty from the beginning
+        last = next(it)
```
Comment (kaushikb11) [inline pytorch_lightning/trainer/supporters.py:523]: How about using `last = next(it, None)`, and if None, return?

Comment (awaelchli) [reply pytorch_lightning/trainer/supporters.py:523]: I didn't know you can do that. But isn't this a trouble when the iterator actually returns None as the real element? Then we can't differentiate

Comment (kaushikb11) [reply pytorch_lightning/trainer/supporters.py:523]: yeah, that's the concern I had as well.

**Review by kaushikb11 — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !

**Pull request merged**